File: formacOS/Tests/AnyClipCoreTests/PeerStateTests.swift
import Testing
@testable import AnyClipCore

@Test func initialIsIdleWithNoPeers() {
    #expect(PeerUIState.initial.kind == .idle)
    #expect(PeerUIState.initial.peers.isEmpty)
    #expect(PeerUIState.initial.sortedPeerNames.isEmpty)
}

@Test func linkUpAddsPeerAndGoesLinked() {
    let s = reducePeerState(.initial, .linkUp(nodeID: "abc", peerName: "win-pc"), now: 42.0)
    #expect(s.kind == .linked)
    #expect(s.peers == ["abc": "win-pc"])
    #expect(s.since == 42.0)
    #expect(s.consecutiveHandshakeFails == 0)
    #expect(s.peerName == "win-pc")            // back-compat accessor
}

@Test func twoPeersRenderSortedNames() {
    var s = reducePeerState(.initial, .linkUp(nodeID: "n2", peerName: "win-pc"), now: 1)
    s = reducePeerState(s, .linkUp(nodeID: "n1", peerName: "android-9"), now: 2)
    #expect(s.kind == .linked)
    #expect(s.peers.count == 2)
    #expect(s.sortedPeerNames == ["android-9", "win-pc"])  // ordinal sort by name
    #expect(s.since == 1)                                  // "since first peer" preserved
}

@Test func linkDownRemovesOnlyThatPeer() {
    var s = reducePeerState(.initial, .linkUp(nodeID: "a", peerName: "p-a"), now: 1)
    s = reducePeerState(s, .linkUp(nodeID: "b", peerName: "p-b"), now: 2)
    s = reducePeerState(s, .linkDown(nodeID: "a", reason: "peer disconnected"), now: 3)
    #expect(s.kind == .linked)                 // b still linked
    #expect(s.peers == ["b": "p-b"])
}

@Test func linkDownOfLastPeerGoesSearching() {
    let linked = reducePeerState(.initial, .linkUp(nodeID: "x", peerName: "p"), now: 1)
    let s = reducePeerState(linked, .linkDown(nodeID: "x", reason: "peer disconnected"), now: 2)
    #expect(s.kind == .searching)
    #expect(s.reason == "peer disconnected")
    #expect(s.peers.isEmpty)
}

@Test func discoveryMovesIdleToSearching() {
    let s = reducePeerState(.initial, .peerDiscovered(name: "n", addr: "1.2.3.4:24816"), now: 1)
    #expect(s.kind == .searching)
}

@Test func discoveryMovesErrorToSearching() {
    let err = reducePeerState(.initial, .permissionMissing(kind: "local_network"), now: 1)
    let s = reducePeerState(err, .peerDiscovered(name: "n", addr: "a"), now: 2)
    #expect(s.kind == .searching)
}

@Test func discoveryDoesNotFlapLinked() {
    let linked = reducePeerState(.initial, .linkUp(nodeID: "x", peerName: "p"), now: 1)
    let s = reducePeerState(linked, .peerDiscovered(name: "n", addr: "a"), now: 2)
    #expect(s == linked)
}

@Test func permissionMissingIsError() {
    let s = reducePeerState(.initial, .permissionMissing(kind: "local_network"), now: 1)
    #expect(s.kind == .error)
    #expect(s.reason == "local_network")
}

@Test func fiveHandshakeFailsTripAuthErrorWhenNoPeers() {
    var s = PeerUIState.initial
    for i in 1...(handshakeFailThreshold - 1) {
        s = reducePeerState(s, .handshakeFailed(addr: "a", reason: "auth"), now: Double(i))
        #expect(s.kind == .idle)
        #expect(s.consecutiveHandshakeFails == i)
    }
    s = reducePeerState(s, .handshakeFailed(addr: "a", reason: "auth"), now: 5)
    #expect(s.kind == .error)
    #expect(s.reason == "auth")
}

@Test func handshakeFailsDoNotTripErrorWhileAPeerIsLinked() {
    var s = reducePeerState(.initial, .linkUp(nodeID: "x", peerName: "p"), now: 1)
    for i in 1...(handshakeFailThreshold + 2) {
        s = reducePeerState(s, .handshakeFailed(addr: "a", reason: "auth"), now: Double(i))
    }
    #expect(s.kind == .linked)                 // an existing link masks the auth escalation
    #expect(s.peers == ["x": "p"])
}

@Test func linkUpResetsFailCounter() {
    var s = PeerUIState.initial
    s = reducePeerState(s, .handshakeFailed(addr: "a", reason: "auth"), now: 1)
    s = reducePeerState(s, .linkUp(nodeID: "x", peerName: "p"), now: 2)
    #expect(s.kind == .linked)
    #expect(s.consecutiveHandshakeFails == 0)
}

File: forwindows/tests/AnyClipCore.Tests/PeerStateTests.cs
using AnyClip.Core;
using Xunit;

namespace AnyClip.Core.Tests;

public class PeerStateTests
{
    [Fact] public void InitialIsIdleWithNoPeers()
    {
        Assert.Equal(PeerStateKind.Idle, PeerUiState.Initial.Kind);
        Assert.Empty(PeerUiState.Initial.Peers);
    }

    [Fact]
    public void LinkUpAddsPeerKeyedByNodeIdAndGoesLinked()
    {
        var s = PeerStateReducer.Reduce(PeerUiState.Initial,
            new LinkUp("node-abc", "win-pc"), 42.0);
        Assert.Equal(PeerStateKind.Linked, s.Kind);
        Assert.Equal("win-pc", s.Peers["node-abc"]);
        Assert.Single(s.Peers);
        Assert.Equal(42.0, s.Since);
        Assert.Equal(0, s.ConsecutiveHandshakeFails);
    }

    [Fact]
    public void SecondLinkUpAddsSecondPeerAndKeepsSince()
    {
        var s = PeerStateReducer.Reduce(PeerUiState.Initial, new LinkUp("a", "mac"), 1);
        s = PeerStateReducer.Reduce(s, new LinkUp("b", "win"), 9);
        Assert.Equal(PeerStateKind.Linked, s.Kind);
        Assert.Equal(2, s.Peers.Count);
        Assert.Equal(1, s.Since); // first-link timestamp retained
    }

    [Fact]
    public void LinkDownRemovesOnlyThatNodeId()
    {
        var s = PeerStateReducer.Reduce(PeerUiState.Initial, new LinkUp("a", "mac"), 1);
        s = PeerStateReducer.Reduce(s, new LinkUp("b", "win"), 2);
        s = PeerStateReducer.Reduce(s, new LinkDown("a", "peer disconnected"), 3);
        Assert.Equal(PeerStateKind.Linked, s.Kind); // still one peer
        Assert.False(s.Peers.ContainsKey("a"));
        Assert.True(s.Peers.ContainsKey("b"));
    }

    [Fact]
    public void LastLinkDownGoesSearchingWithReason()
    {
        var s = PeerStateReducer.Reduce(PeerUiState.Initial, new LinkUp("a", "mac"), 1);
        s = PeerStateReducer.Reduce(s, new LinkDown("a", "peer disconnected"), 2);
        Assert.Equal(PeerStateKind.Searching, s.Kind);
        Assert.Empty(s.Peers);
        Assert.Equal("peer disconnected", s.Reason);
    }

    [Fact]
    public void UnknownLinkDownIsANoOp()
    {
        var s = PeerStateReducer.Reduce(PeerUiState.Initial, new LinkUp("a", "mac"), 1);
        var s2 = PeerStateReducer.Reduce(s, new LinkDown("ghost", "x"), 2);
        Assert.Same(s.Peers, s2.Peers); // untouched -> same reference
        Assert.Equal(PeerStateKind.Linked, s2.Kind);
    }

    [Fact]
    public void DiscoveryMovesIdleAndErrorToSearchingOnlyWhenNoPeers()
    {
        Assert.Equal(PeerStateKind.Searching,
            PeerStateReducer.Reduce(PeerUiState.Initial, new PeerDiscovered("n", "a"), 1).Kind);
        var err = PeerStateReducer.Reduce(PeerUiState.Initial, new PermissionMissing("x"), 1);
        Assert.Equal(PeerStateKind.Searching,
            PeerStateReducer.Reduce(err, new PeerDiscovered("n", "a"), 2).Kind);
        var linked = PeerStateReducer.Reduce(PeerUiState.Initial, new LinkUp("a", "p"), 1);
        Assert.Same(linked, PeerStateReducer.Reduce(linked, new PeerDiscovered("n", "a"), 2));
    }

    [Fact]
    public void FiveHandshakeFailsTripAuthErrorFromIdle()
    {
        var s = PeerUiState.Initial;
        for (int i = 1; i < PeerStateReducer.HandshakeFailThreshold; i++)
        {
            s = PeerStateReducer.Reduce(s, new HandshakeFailed("a", "auth"), i);
            Assert.Equal(PeerStateKind.Idle, s.Kind);
            Assert.Equal(i, s.ConsecutiveHandshakeFails);
        }
        s = PeerStateReducer.Reduce(s, new HandshakeFailed("a", "auth"), 5);
        Assert.Equal(PeerStateKind.Error, s.Kind);
        Assert.Equal("auth", s.Reason);
        Assert.Equal(5, s.ConsecutiveHandshakeFails);
    }

    [Fact]
    public void LinkUpResetsFailCounter()
    {
        var s = PeerStateReducer.Reduce(PeerUiState.Initial, new HandshakeFailed("a", "auth"), 1);
        s = PeerStateReducer.Reduce(s, new LinkUp("a", "p"), 2);
        Assert.Equal(PeerStateKind.Linked, s.Kind);
        Assert.Equal(0, s.ConsecutiveHandshakeFails);
        s = PeerStateReducer.Reduce(s, new HandshakeFailed("a", "auth"), 3);
        Assert.NotEqual(PeerStateKind.Error, s.Kind); // still linked, one fail
        Assert.Equal(1, s.ConsecutiveHandshakeFails);
    }

    [Fact] public void ThresholdConstantIsFive() =>
        Assert.Equal(5, PeerStateReducer.HandshakeFailThreshold);

    [Fact]
    public void TrayIconSpecMapping()
    {
        var linked = PeerStateReducer.Reduce(PeerUiState.Initial, new LinkUp("a", "p"), 1);
        Assert.Equal(new TrayIconSpec(false, false), TrayIconSpec.For(linked));
        Assert.Equal(new TrayIconSpec(true, false), TrayIconSpec.For(PeerUiState.Initial));
        var searching = PeerStateReducer.Reduce(PeerUiState.Initial, new PeerDiscovered("n", "a"), 1);
        Assert.Equal(new TrayIconSpec(true, false), TrayIconSpec.For(searching));
        var err = PeerStateReducer.Reduce(PeerUiState.Initial, new PermissionMissing("x"), 1);
        Assert.Equal(new TrayIconSpec(true, true), TrayIconSpec.For(err));
    }
}
